### PR TITLE
[feat] normalise column names in `read_csv`

### PR DESCRIPTION
**Pull request by MattBuckley-TfN**
# Changes

Optional flag in `read_csv` to normalise column names when reading, normalising does the following:
- Strips whitespace from the ends and converts to lower case
- Replaces spaces with underscores; and
- Removes any characters not in this `abcdefghijklmnopqrstuvwxyz0123456789!#$%£&-.<=>+^_~()`

# Tasks

- [x] Have unittests been added (testing individual functions and their edge cases)
- [x] Has documentation (including user guide) been updated
- [x] Have new dependencies been added (or changed) in relevant `requirements.txt` - No
- [x] Code linting, tests and other workflows pass

----
📚 Documentation preview 📚: https://caftoolkit--199.org.readthedocs.build/en/199/

### CODE REVIEW

File: src/caf/toolkit/io.py [hunk 1]
```diff
@@ -114,39 +126,175 @@ def read_csv(path: os.PathLike, name: str | None = None, **kwargs) -> pd.DataFra
     if not path.is_file():
         raise FileNotFoundError(f"{name} file does not exist: '{path}'")
 
+    column_lookup = None
+    if normalise_column_names:
+        column_lookup, *parameters = _normalise_read_csv(path, **kwargs)
```
Comment (Kieran-Fishwick-TfN) [inline src/caf/toolkit/io.py:131]: I find this a bit confusing.
What is the thinking behind, reading one row to normalise the columns before reading in the whole thing.
Maybe add some quick comments?

File: src/caf/toolkit/io.py [hunk 2]
```diff
@@ -114,39 +126,175 @@ def read_csv(path: os.PathLike, name: str | None = None, **kwargs) -> pd.DataFra
...
+    duplicates = collections.defaultdict(list)
+    for col in original:
+        normalised = _normalise_name(col)
+
+        if normalised in lookup.values():
+            duplicates[normalised].append(col)
+
+        lookup[col] = normalised
+
+    if len(duplicates) > 0:
```
Comment (Kieran-Fishwick-TfN) [inline src/caf/toolkit/io.py:253]: Could follow pandas protocol and add numbers?

**Review by Kieran-Fishwick-TfN — APPROVED** [review]
Looks good, happy to be merged as is. Just added some comments to consider

**Review by Kieran-Fishwick-TfN — APPROVED** [review]
